core/modules/replica/Configuration.cc

```diff
@@ -23,818 +23,1732 @@
+// Instance API of class Configuration
+// ------------------------------------------------------------------------------
+
+Configuration::Configuration()
+    :   _data(ConfigurationSchema::defaultConfigData()) {
 }
 
 
-string Configuration::databaseTechnology() const {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->databaseTechnology();
+void Configuration::reload() {
+    util::Lock const lock(_mtx, _context(__func__));
+    if (!_configUrl.empty()) {
+        bool const reset = true;
+        _load(lock, _configUrl, reset);
+    }
 }
 
 
-string Configuration::databaseHost() const {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->databaseHost();
+void Configuration::reload(string const& configUrl) {
+    util::Lock const lock(_mtx, _context(__func__));
+    bool const reset = true;
+    _load(lock, configUrl, reset);
 }
 
 
-uint16_t Configuration::databasePort() const {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->databasePort();
+void Configuration::reload(json const& obj) {
+    util::Lock const lock(_mtx, _context(__func__));
+    bool const reset = true;
+    _load(lock, obj, reset);
 }
 
 
-string Configuration::databaseUser() const {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->databaseUser();
+string Configuration::configUrl(bool showPassword) const {
+    util::Lock const lock(_mtx, _context(__func__));
+    if (_connectionPtr == nullptr) return string();
+    return _connectionParams.toString(showPassword);
 }
 
 
-string Configuration::databasePassword() const {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->databasePassword();
+map<string, set<string>> Configuration::parameters() const {
+    return ConfigurationSchema::parameters();
 }
 
 
-string Configuration::databaseName() const {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->databaseName();
+string Configuration::getAsString(string const& category, string const& param) const {
+    util::Lock const lock(_mtx, _context(__func__));
+    return ConfigurationSchema::json2string(
+            _context(__func__) + " category: '" + category + "' param: '" + param + "' ",
+            _get(lock, category, param));
 }
 
 
-size_t Configuration::databaseServicesPoolSize() const {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->databaseServicesPoolSize();
+void Configuration::setFromString(string const& category, string const& param,
+                                  string const& val, bool updatePersistentState) {
+    json obj;
+    {
+        util::Lock const lock(_mtx, _context(__func__));
+        obj = _get(lock, category, param);
+    }
+    if (obj.is_string()) {
+        Configuration::set<string>(category, param, val, updatePersistentState);
+    } else if (obj.is_number_unsigned()) {
+        Configuration::set<uint64_t>(category, param, stoull(val), updatePersistentState);
```

Comment:
What if parameter type is `uint_16t`, will overflow be detected at some point?

Reply:
Unfortunately, there is some non-determinism when it comes to numeric types in the JSON library. It won't tell you the precision of a type. I have no simple solution for that. There is a possibility of the numeric overflow if some large number would be provided where `uint16_t` (or `uint32_t`) is expected.